zkevm-circuits/src/evm_circuit/util/math_gadget/cmp_words.rs

```diff
@@ -1,48 +1,40 @@
+use std::marker::PhantomData;
+
 use crate::{
     evm_circuit::util::{
-        self, constraint_builder::EVMConstraintBuilder, from_bytes, math_gadget::*, select,
-        CachedRegion,
+        constraint_builder::EVMConstraintBuilder, from_bytes, math_gadget::*, select, CachedRegion,
     },
-    util::Expr,
+    util::{word::WordExpr, Expr},
 };
 use eth_types::{Field, ToLittleEndian, Word};
 use halo2_proofs::plonk::{Error, Expression};
 
 #[derive(Clone, Debug)]
 /// CmpWordsGadget compares two words, exposing `eq`  and `lt`
-pub(crate) struct CmpWordsGadget<F> {
+pub(crate) struct CmpWordsGadget<F, T> {
     comparison_lo: ComparisonGadget<F, 16>,
     comparison_hi: ComparisonGadget<F, 16>,
     pub eq: Expression<F>,
     pub lt: Expression<F>,
+    marker: PhantomData<T>,
 }
 
-impl<F: Field> CmpWordsGadget<F> {
-    pub(crate) fn construct(
-        cb: &mut EVMConstraintBuilder<F>,
-        a: &util::Word<F>,
-        b: &util::Word<F>,
-    ) -> Self {
-        // `a[0..16] <= b[0..16]`
-        let comparison_lo = ComparisonGadget::construct(
-            cb,
-            from_bytes::expr(&a.cells[0..16]),
-            from_bytes::expr(&b.cells[0..16]),
-        );
+impl<F: Field, T: WordExpr<F>> CmpWordsGadget<F, T> {
+    pub(crate) fn construct(cb: &mut EVMConstraintBuilder<F>, a: T, b: T) -> Self {
```

Comment:
Consider having two types (one for `a` and one for `b`).  In theory we don't need them to be the same, we just need to be able to get the `lo/hi` of each one.

Reply:
This is really nice call!! Type `a, b` is separated